raster-foundry/raster-foundry
# Add annotation project post endpoint

## Overview

Brief description of what this PR does, and why it is needed.

### Checklist

- ~Swagger specification updated~ explicitly skipping since we're not prioritizing API users

### Notes

~There's a bit of a mismatch here in what the `ObjectPermissions` trait expects (`owner`) and therefore what we need for sharing and how the migration was originally written to migrate existing annotation projects. I didn't resolve it here because I'm confused about timelines. I _think_ it's the case that that job needs to be updated to insert an `owner` field instead of a `created_by` field though.~ verified with bbdce40

I didn't add a new domain or scoped actions. This is just project creation, and I didn't want to think through the universe where someone has a combined limit on RF + Groundwork projects.

Methods for `ObjectPermissions` are just stubs right now -- routes are part of scope for #5282

## Testing Instructions

- give your user `groundworkUser` scope in the db
- Save this to `annot-project.json`:

```json
{
    "name": "new-project",
    "projectType": "CLASSIFICATION",
    "tileLayers": [],
    "labelClassGroups": []
}
```

- assemble server, bring up frontend, get a token...
- `export JWT_AUTH_TOKEN=<your token without Bearer prefix>`
- `cat annot-project.json | http --auth-type=jwt :9091/api/annotation-projects`
- it should succeed
- add some more exciting stuff:
  - put a few things like this in `tileLayers`:

```json
	{
	    "name": "tiles!",
	    "url": "https://tiles.omg/{z}/{x}/{y}",
	    "layerType": "TMS"
	}
```

  - put a few things like this in `labelClassGroups`:

```json
	{
	    "name": "green things",
	    "classes": [
		{
		    "name": "so green",
		    "colorHexCode": "#00ff00",
		    "index": 0
		},
		{
		    "name": "pretty green",
		    "colorHexCode": "#00dd00",
		    "index": 0
		}
	    ]
	}
```

- `cat annot-project.json | http --auth-type=jwt :9091/api/annotation-projects` again
- it should succeed again
- keep doing this until you get a 403 -- that's from the scope limit
- then reassemble your batch jar and reload development data for consistency
- `./scripts/console batch 'java -cp /opt/raster-foundry/jars/batch-assembly.jar com.rasterfoundry.batch.Main liberate-annotation-projects http://localhost:8081'` -- the project migration job should still work. This tests that the project migration works _after_ application of the migrations here. We might have to do some fancy flyway magic to make everything play nicely once #5289 is complete

Closes #5283
